draft-barnes-mls-appsync.md

```diff
@@ -60,171 +61,225 @@ communicate. Applications based on MLS can integrate their state into this
-are reflected in the group's `application_states`.
-
-> **NOTE:** This design exposes the high-level structure of the application state
-> to MLS.  An alternative design would be to have the application state be opaque
-> to MLS.  There is a trade-off between generality and the complexity of the API
-> between the MLS implementation and the application.  An opaque design would give
-> the application more freedom, but require the MLS stack to call out to the
-> application to get the updated state as part of Commit processing.  This design
-> allows the updates to happen within the MLS stack, so that no callback is
-> needed, at the cost of forcing the application state to fit a certain structure.
-> It also potentially can result in smaller state updates in large groups.
-
-The state for Each `applicationId` in the `application_states` needs to conform
-to one of four basic types: an ordered array, an unordered array, a map, or an
-irreducible blob. This allows the AppSync proposal to efficiently modify a large
-application state object.
-
-The content of the `application_states` extension and the `AppSync` proposal are
-structured as follows:
+This document defines a new GroupContextExtensionsDiff proposal. It
+is not a Safe Extension as defined in {{Section 2 of
+!I-D.ietf-mls-extensions}}, because it can modify any GroupContext
+extension, including those defined in {{!RFC9420}}. It is intended as a
+complete replacement for the GroupContextExtensions Proposal.
+The GroupContextExtensionsDiff proposal does not require an UpdatePath,
+and may be sent by an authorized external sender.
 
 ~~~ tls
 enum {
-    irreducible(0),
-    map(1),
-    unorderedList(2),
-    orderedArray(3),
+    remove(0),
+    add(1),
+    replace(2),
+    diff(3),
     (255)
-} StateType;
+} OperationType;
+
+uint8 DiffType;
 
+struct {
+    ExtensionType group_context_extension;
+    OperationType operation;
+    select (operation) {
+        case remove:
+            struct {};
+        case add:
+            opaque extension_data<V>;
+        case replace:
+            opaque extension_data<V>;
+        case diff:
+            opaque diff_data<V>;
+    };
+} ExtensionDiff;
+
+struct {
+    ExtensionDiff group_context_extension_diffs<V>;
+} GroupContextExtensionsDiff;
+~~~
+
+The semantics of OperationType are as follows:
+
+- `remove` means that the GroupContext extension was present, and will be
+completely absent from the GroupContext.
+- `add` means that the GroupContext extension was absent, and will be
+present with the value provided in `extension_data`.
+- `replace` means that the GroupContext extension was present and will
+be completely replaced with the new value in `extension_data`
+- `diff` means that the GroupContext extension will be modified according
+to the diff format defined by that extension. (Two concrete diff formats are
+defined in {{diff-formats}} which extension designers are free to use or
+not use.)
+
+
+## Processing rules
```

Comment:
This doesn't need to be a subsection.  I would follow the same structure as [GCE](https://rfcs.online/rfcs/rfc9420.html#name-groupcontextextensions), "A member of the group applies a GroupContextExtensionDiff proposal with the following steps:".  That can incorporate the OperationType discussion above as well.